Integrate questionary

akkuldn opened the pull request:
This is a new feature that allows the user to run the tests interactively via the terminal.
__Steps to run:__
1) If you're using Gitbash for windows please make sure to set your bash alias by adding the following command to your `bash_rc` file `alias python='winpty python.exe'`
2) Install the requirements using `pip install -r reqquirements.txt`
3) Commands to run the test : 
`python -m pytest tests/test_example_form.py --interactivemode_flag Y`
`python -m pytest tests/test_mobile_bitcoin_price.py --interactivemode_flag Y`
` python -m pytest tests/test_api_example.py --interactivemode_flag Y`

Changes made:
---
- A new feature to run the tests interactive using `--interactivemode_flag`
- A new pytest fixture `test_api_obj` in contest file for API tests
- Importing the API URL from a conf file

qxf2 commented on conftest.py:
```diff
@@ -71,6 +89,23 @@ def test_mobile_obj(mobile_os_name, mobile_os_version, device_name, app_package,
         print("Python says:%s"%str(e))
 
+@pytest.fixture
+def test_api_obj(interactivemode_flag,api_url=api_example_conf.api_url):
```
Why does the new argument not have a default?

qxf2 commented on conftest.py:
```diff
@@ -3,14 +3,28 @@
     try:
+
+        if interactivemode_flag.lower() == "y":
+
```
The whiteline spacing here is off. Why are there empty lines after the if?

akkuldn replied:
Fixed

qxf2 commented on endpoints/API_Player.py:
```diff
@@ -10,17 +10,30 @@
 import logging
 from conf import api_example_conf as conf
+from utils import interactive_mode
+#from conftest import interactivemode_flag
```
How are commented out lines of code slipping through to a PR?

akkuldn replied:
Fixed

qxf2 commented on utils/interactive_mode.py:
```diff
@@ -0,0 +1,421 @@
+"""
```
@akkuldn - I feel like this file will be extremely hard to maintain. A few suggestions:
a) Break your methods down better. Typically, methods should not be more than 10 lines. You should break the code down better.
b) When you intersperse code and strings, the strings should become configurable. If you are using the same strings in two or more places, you should be making it a variable. that way, when the hardcoded string needs to change, the QA changes them in only one spot.
c) Can you please add some unit tests for these options? Include them in the PR to help your reviewer. You can delete the tests once they approve your PR.

akkuldn replied:
a) Brokedown the methods making them smaller
b) Now having a single variable for most of the reused strings
c) Added some unit tests
Fixed

qxf2 reviewed (changes requested):
@akkuldn - I got a little time to check out your work. This is a very good addition to our framework. 

I have left some comments based on ~5-minutes of code review. Please consider them along with the feedback you get from your other reviewers.

PreedhiVivek commented:
@akkuldn 
An interesting and useful feature add-on to our framework. Nice Akkul!
My observations:
- Errors out while selecting the firefox version
![browser_version_error](https://user-images.githubusercontent.com/52440868/114386783-4c807080-9baf-11eb-944b-c94c4893140b.PNG)

- Do you want to run the test with a default set of options? No -> What would you like to change? Exit -> Errors out. Would be good to go back to the default options question on Exit. 
![error_on_exit](https://user-images.githubusercontent.com/52440868/114386814-56a26f00-9baf-11eb-92e7-d491cc3d4cd9.PNG)

- ask_questions_<>() functions will have to be unit tested well, while modifying different default options and ensuring expected behaviour.

- In Readme.md '(This will give the user a menu of different choices allowing the user to pick their desired configuration to run the test)' can be rephrased to (This option will allow the user to pick the desired configuration to run the test, from the menu displayed)' and typo `aliash`.  Is that __Note__ even required? Isn't that understood and holds good for running the test suit with any other option as well.  You can probably make it a common note under that section?

- conftest.py Watch out for line spacing. Could see many empty lines and very long lines of code. I suggest, running your modified files against pylint and radon. You could fix the ones that belong to your change, if all is too much or parked as a separate task.

- API_Player.py remove commented out code line # 14, 26, 27  

- Interactive_mode.py, functions such as ask_questions_gui() and ask_questions_mobile() has be broken down. They are too long. Also, ensure this new module is pylint and radon compliant.

- Commits will have to be squashed before merge.  

- I will be running the API tests and will update if I find any comments.

avinash010 commented on Readme.md:
```diff
@@ -150,6 +150,10 @@ a)py.test [options]
 	--tesults 	used to report test results to tesults			E.g: python -m pytest test_example_form.py --tesults Y(This will report test report to tesults)
 
+	--interactivemode_flag	used to run the tests interactively __Note:__ If your running the test on git bash for windows please make sure to set your bash aliash by adding the following command to bash_rc `alias python='winpty python.exe'`         E.g:  python -m pytest tests/test_example_form.py --interactivemode_flag Y(This will give the user a menu of different choices allowing the user to pick  ...
+
```
There are empty lines here too.

akkuldn replied:
Fixed

avinash010 commented on Readme.md:
```diff
@@ -150,6 +150,10 @@ a)py.test [options]
 	-n 	used to run tests in parallel					E.g: python -m pytest -n 3 -v (This will run three tests in parallel)
 	--tesults 	used to report test results to tesults			E.g: python -m pytest test_example_form.py --tesults Y(This will report test report to tesults)
 
+	--interactivemode_flag	used to run the tests interactively __Note:__ If your running the test on git bash for windows please make sure to set your bash aliash by adding the following command to bash_rc `alias python='winpty python.exe'`         E.g:  python -m pytest tests/test_example_form.py --interactivemode_flag Y(This will give the user a menu of different choices allowing the user to pick  ...
```
Shouldn't this be named interactive_mode_flag as per the naming convention we follow?

akkuldn replied:
Fixed

akkuldn commented:
> @akkuldn
> An interesting and useful feature add-on to our framework. Nice Akkul!
> My observations:
> 
> * Errors out while selecting the firefox version, when I want to change default options:
>   ![browser_version_error](https://user-images.githubusercontent.com/52440868/114386783-4c807080-9baf-11eb-944b-c94c4893140b.PNG)
> * Do you want to run the test with a default set of options? No -> What would you like to change? Exit -> Errors out. Would be good to go back to the default options question on Exit.
>   ![error_on_exit](https://user-images.githubusercontent.com/52440868/114386814-56a26f00-9baf-11eb-92e7-d491cc3d4cd9.PNG)
> * ask_questions_<>() functions will have to be unit tested well, while modifying different default options and ensuring expected behaviour.
> * In Readme.md '(This will give the user a menu of different choices allowing the user to pick their desired configuration to run the test)' can be rephrased to (This option will allow the user to pick the desired configuration to run the test, from the menu displayed)' and typo `aliash`.  Is that **Note** even required? Isn't that understood and holds good for running the test suit with any other option as well.  You can probably make it a common note under that section?
> * conftest.py Watch out for line spacing. Could see many empty lines and very long lines of code. I suggest, running your modified files against pylint and radon. You could fix the ones that belong to your change, if all is too much or parked as a separate task.
> * API_Player.py remove commented out code line # 14, 26, 27
> * Interactive_mode.py, functions such as ask_questions_gui() and ask_questions_mobile() has be broken down. They are too long. Also, ensure this new module is pylint and radon compliant.
> * Commits will have to be squashed before merge.
> * I will be running through the API test changes and will update if I find any comments.

---
Thanks for the feedback Preedhi,
1) Fixed
2) I added the exit option intentionally so that the user could exit out of the while loop incase he doesn't want to run the test
3) Added unit tests, you can find it inside the `tests` directory `test_interactive_mode.py`
4) Fixed
5) Fixed
6) Fixed
7) Fixed
8) Ill squash the commits once  I get the approval to merge

Please do let me know of any further comments

PreedhiVivek commented:
@akkuldn 
My re-reviewing comments:
Readme file:
- Remove empty line # 152,  _Note:_ to NOTE: can be changed to be in sync with existing convention and would be good to have `interactive_mode` as the optional command line parameter in-place of `interactive_mode_flag`

conftest.py
- unwanted print line # 97

> > * Errors out while selecting the firefox version, when I want to change default options:
> >   ![browser_version_error](https://user-images.githubusercontent.com/52440868/114386783-4c807080-9baf-11eb-944b-c94c4893140b.PNG)
Test runs smoothly if I change browser to Firefox. However, when I tried Firefox+Windows+10 combination, observed that
Firefox browser launches and the cursor keeps blinking at the address bar. In the console,
`Errros out when I select Firefox, Windows, version 10
Exception when trying to run test: C:\Users\PreedhiVivek\code\POM\qxf2-page-object-model\conftest.py
Python says:Message: permission denied`
> > * ask_questions_<>() functions will have to be unit tested well, while modifying different default options and ensuring expected behaviour.
Unit tests should ensure all such combinations are covered. Please ensure that.

> > * Do you want to run the test with a default set of options? No -> What would you like to change? Exit -> Errors out. Would be good to go back to the default options question on Exit.
> >   ![error_on_exit](https://user-images.githubusercontent.com/52440868/114386814-56a26f00-9baf-11eb-92e7-d491cc3d4cd9.PNG)
I understand the use-case Akkul. But the handling of this case should be looked into. It currently errors out when Exit is chosen from the menu. Maybe, catch the exception and log a good message.

> > * Interactive_mode.py, functions such as ask_questions_gui() and ask_questions_mobile() has be broken down. They are too long. Also, ensure this new module is pylint and radon compliant.
I do see that, you have broken down code well now. However, it has to be done a bit more. Pylint of this module is rated at 7.72/10 and radon run also points out few functions rated B/C which need to be addressed.

akkuldn commented:
> @akkuldn
> My re-reviewing comments:
> Readme file:
> 
> * Remove empty line # 152,  `_Note:_` to NOTE: can be changed to be in sync with existing convention and would be good to have `interactive_mode` as the optional command line parameter in-place of `interactive_mode_flag`
> 
> conftest.py
> 
> * unwanted print line # 97
> 
> > > * Errors out while selecting the firefox version, when I want to change default options:
> > >   ![browser_version_error](https://user-images.githubusercontent.com/52440868/114386783-4c807080-9baf-11eb-944b-c94c4893140b.PNG)
> 
> Test runs smoothly if I change browser to Firefox. However, when I tried Firefox+Windows+10 combination, observed that
> Firefox browser launches and the cursor keeps blinking at the address bar. In the console,
> `Errros out when I select Firefox, Windows, version 10 Exception when trying to run test: C:\Users\PreedhiVivek\code\POM\qxf2-page-object-model\conftest.py Python says:Message: permission denied`
> 
> > > * ask_questions_<>() functions will have to be unit tested well, while modifying different default options and ensuring expected behaviour.
> 
> Unit tests should ensure all such combinations are covered. Please ensure that.
> 
> > > * Do you want to run the test with a default set of options? No -> What would you like to change? Exit -> Errors out. Would be good to go back to the default options question on Exit.
> > >   ![error_on_exit](https://user-images.githubusercontent.com/52440868/114386814-56a26f00-9baf-11eb-92e7-d491cc3d4cd9.PNG)
> 
> I understand the use-case now Akkul. But the handling of this case should be looked into. It currently errors out when Exit is chosen from the menu. Maybe, catch the exception and log a good message.
> 
> > > * Interactive_mode.py, functions such as ask_questions_gui() and ask_questions_mobile() has be broken down. They are too long. Also, ensure this new module is pylint and radon compliant.
> 
> I do see that, you have broken down code well now. However, it has to be done a bit more. Pylint of this module is rated at 7.72/10 and radon run also points out few functions rated B/C which need to be addressed.

@PreedhiVivek ,
1) Removed line and changed _Note_ to note
2) removed unwanted print
3) I could not replicate the issue. But however, os_name, browser_versions, Os_version is used only when  running the test remotely. So i have made changes to the code such that you wont be able to select the OS Name or browser version unless your remote_flag is not set to "Yes". And I cant add more unit test scenarios as it requires human interaction with terminal for each test, and I gotta change the test file for every test case manually
4) Added message when exiting
5) Improved the code, now getting pyling rating of 9.05, now there are  four functions that show b,c rating in radon, i can't break those functions any smaller as they are nested if statements

PreedhiVivek commented:
@akkuldn I am good with your changes.
One final formatting suggestion in Readme,
![ReadMe_currently](https://user-images.githubusercontent.com/52440868/120285850-4a977b80-c2db-11eb-8b11-1187f58b281c.PNG)
To improve readability,
![ReadMe_spacing](https://user-images.githubusercontent.com/52440868/120285906-5b47f180-c2db-11eb-853a-c04e7b810360.PNG)

akkuldn commented:
> @akkuldn I am good with your changes.
> One final formatting suggestion in Readme,
> ![ReadMe_currently](https://user-images.githubusercontent.com/52440868/120285850-4a977b80-c2db-11eb-8b11-1187f58b281c.PNG)
> To improve readability,
> ![ReadMe_spacing](https://user-images.githubusercontent.com/52440868/120285906-5b47f180-c2db-11eb-853a-c04e7b810360.PNG)

Made the change @PreedhiVivek
